src/container/srv_target.c

```diff
@@ -151,18 +276,102 @@ cont_stop_agg_ult(struct ds_cont_child *cont)
 
 	while (cont->sc_vos_aggregating)
 		ABT_thread_yield();
+}
+
+/* ds_cont_child *******************************************************/
+
+static inline struct ds_cont_child *
+cont_child_obj(struct daos_llink *llink)
+{
+	return container_of(llink, struct ds_cont_child, sc_list);
+}
+
+static int
+cont_child_alloc_ref(void *co_uuid, unsigned int ksize, void *po_uuid,
+		     struct daos_llink **link)
+{
+	struct ds_cont_child	*cont;
+	int			rc;
+
+	D_ASSERT(po_uuid != NULL);
+	D_DEBUG(DF_DSMS, DF_CONT": opening\n", DP_CONT(po_uuid, co_uuid));
+
+	D_ALLOC_PTR(cont);
+	if (cont == NULL)
+		return -DER_NOMEM;
+
+	rc = ABT_mutex_create(&cont->sc_mutex);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+
+	rc = ABT_cond_create(&cont->sc_dtx_resync_cond);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out_mutex;
+	}
 
-	D_DEBUG(DF_DSMS, DF_CONT": Aggregation ULT stopped\n",
-		DP_CONT(NULL, cont->sc_uuid));
+	cont->sc_pool = ds_pool_child_lookup(po_uuid);
+	if (cont->sc_pool == NULL) {
+		rc = -DER_NO_HDL;
+		goto out_cond;
+	}
+
+	rc = vos_cont_open(cont->sc_pool->spc_hdl, co_uuid, &cont->sc_hdl);
+	if (rc != 0)
+		goto out_pool;
+
+	uuid_copy(cont->sc_uuid, co_uuid);
+	cont->sc_dtx_resync_time = crt_hlc_get();
+	cont->sc_aggregation_min = DAOS_EPOCH_MAX;
+	/* prevent aggregation till snapshot iv refreshed */
+	cont->sc_aggregation_max = 0;
+	cont->sc_snapshots_nr = 0;
+	cont->sc_snapshots = NULL;
+	cont->sc_vos_aggregating = 0;
+	cont->sc_abort_vos_aggregating = 0;
+
+	rc = cont_start_agg_ult(cont);
```

Comment:
What's the benefit of start aggression ULT when establish the ds_cont_child instead of opening the container? That may be triggered by non-client behavior, such as rebuild. It seems unnecessary (or may be bring race) to start aggregation under such cases.

Reply:
Then is it expected that the aggregation is triggered for rebuild case?

Reply:
Whether aggregation is triggered on rebuild target container is another story, I think it's unrelated to this change.

My point is that these server background jobs like aggregation or batched commit, they don't need open handle to acquire access right, and they don't have to be stopped when the container is closed by client, right?